langserver/symbol.go

```diff
@@ -239,17 +250,84 @@ func score(q Query, s lsp.SymbolInformation) (scor int) {
 
 // toSym returns a SymbolInformation value derived from values we get
 // from the Go parser and doc packages.
-func toSym(name, container string, kind lsp.SymbolKind, fs *token.FileSet, pos token.Pos) lsp.SymbolInformation {
-	container = filepath.Base(container)
-	if i := strings.LastIndex(container, " "); i >= 0 {
-		container = container[i+1:]
+func toSym(name string, bpkg *build.Package, recv string, kind lsp.SymbolKind, fs *token.FileSet, pos token.Pos) symbolPair {
+	container := recv
+	if container == "" {
+		container = filepath.Base(bpkg.ImportPath)
+	}
+	return symbolPair{
+		SymbolInformation: lsp.SymbolInformation{
+			Name:          name,
+			Kind:          kind,
+			Location:      goRangeToLSPLocation(fs, pos, pos+token.Pos(len(name))-1),
+			ContainerName: container,
+		},
+		// NOTE: fields must be kept in sync with workspace_refs.go:defSymbolDescriptor
+		desc: lspext.SymbolDescriptor{
+			"vendor":      IsVendorDir(bpkg.Dir),
+			"package":     path.Clean(bpkg.ImportPath),
+			"packageName": bpkg.Name,
+			"recv":        recv,
+			"name":        name,
+			"id":          fmt.Sprintf("%s:%s:%s:%s", path.Clean(bpkg.ImportPath), bpkg.Name, recv, name),
+		},
+	}
+}
+
+func parseID(v string) (pkg, packageName, recv, name string) {
```

Comment:
So here is the use case we need to solve - at the moment for Go, soon for as many languages as possible:
- We **have** a unique, human-readable symbol name (referred to as symbol ID)
- We **need** symbol information about the symbol it refers to (location to be precise)

To be unique, the symbol ID will be composed of different parts, in **all languages**.
What parts exactly need to be encoded inside the string and how they are separated the best is **language-specific**.

So, somewhere in the the whole process that solves above use case, the symbol ID needs to be decomposed into its language-specific parts. This requires language-specific code. There are two options:

1. Let the frontend parse the symbol ID (from the URL) in a huge `switch` with language-specific patterns and build a `SymbolDescriptor` out of the part, then query by that.
2. Let the frontend just query by the symbol ID directly, e.g. pass it directly from the URL into a `SymbolDescriptor` that looks like `{ id: idFromUrl }` and query by that. The _language server_ will parse the ID and in the language-specific way. _The language-specific code will stay inside the language servers._

We want 2) because we want to _avoid_ language-specific code outside of language servers, because they should abstract away language specificness by speaking the common protocol that is LSP.

Does that make sense?